pkg/iac/scanners/terraformplan/snapshot/snapshot.go

```diff
@@ -0,0 +1,197 @@
+package snapshot
+
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/liamg/memoryfs"
+)
+
+const (
+	configSnapshotPrefix       = "tfconfig/"
+	configSnapshotManifestFile = configSnapshotPrefix + "modules.json"
+	configSnapshotModulePrefix = configSnapshotPrefix + "m-"
+
+	tfplanFilename = "tfplan"
+)
+
+type (
+	configSnapshotModuleRecord struct {
+		Key        string `json:"Key"`
+		SourceAddr string `json:"Source,omitempty"`
+		Dir        string `json:"Dir"`
+	}
+
+	configSnapshotModuleManifest []configSnapshotModuleRecord
+)
+
+func IsPlanSnapshot(r io.Reader) bool {
+	if r == nil {
+		return false
+	}
+
+	buf, err := io.ReadAll(r)
+	if err != nil {
+		return false
+	}
+
+	zr, err := zip.NewReader(bytes.NewReader(buf), int64(len(buf)))
```

Comment:
It seems we need to know the size as `zip.NewReader` expects it, what if we use [`NewReadSeekerAt()`](https://github.com/aquasecurity/trivy/blob/main/pkg/x/io/io.go#L20)  and get the size via something like the following?

```suggestion
       size, _ := io.Copy(io.Discard, r)
	f, _ := iox.NewReadSeekerAt(r)
	zr, err := zip.NewReader(f, size)
	if err != nil {
		return false
	}
```


Just trying to see if we can avoid reading the entire plan snapshot into memory.

Reply:
Another idea is if we can write a new function that like the `NewReadSeekerAt()` but also returns the size as well.

Reply:
[NewReadSeekerAt](https://github.com/aquasecurity/trivy/blob/main/pkg/x/io/io.go#L20) copies the data from `io.Reader` to a buffer, so there is no way to avoid reading the file completely, or am I wrong?

Reply:
Thanks for the research!  I found an even more effective way https://github.com/aquasecurity/trivy/pull/6176/commits/557177b20a8aacc96a7e2b9e38ad8075389cbf0d